### ISSUE TITLE
Differences between paper and code

### ISSUE THREAD

**Issue by oliviertrottier** [ehoogeboom/e3_diffusion_for_molecules#30]
Hi,
I have been very interested in your equivariant neural network for generating QM9-like molecules and I have been trying to reproduce your results for the unconditional generation without success. I ran the default configurations for training on QM9 and I have observed a few differences between what is said in the paper and what is implemented in the code with the default configs.

Here is what I have identified so far:
1) The node features that are given as input to phi_x at layer l is not **h**<sup>l</sup> but **h**<sup>l+1</sup>.
The `gcl_equiv` module of the `EquivariantBlock` class in `egnn_new.py` essentially implements the phi_x function by instantiating the `EquivariantUpdate` class. As shown below in the `forward()` method of `EquivariantBlock`, the `gcl_equiv` module takes the **output** `h` since it's been overwritten before giving it as input. I would define two variables (`h_in` and `h_out`) to distinguish them.
```
    def forward(self, h, x, edge_index, node_mask=None, edge_mask=None, edge_attr=None):
        # Edit Emiel: Remove velocity as input
        distances, coord_diff = coord2diff(x, edge_index, self.norm_constant)
        if self.sin_embedding is not None:
            distances = self.sin_embedding(distances)
        edge_attr = torch.cat([distances, edge_attr], dim=1)
        for i in range(0, self.n_layers):
            h, _ = self._modules["gcl_%d" % i](h, edge_index, edge_attr=edge_attr, node_mask=node_mask, edge_mask=edge_mask)
        x = self._modules["gcl_equiv"](h, x, edge_index, coord_diff, edge_attr, node_mask, edge_mask)

        # Important, the bias of the last linear might be non-zero
        if node_mask is not None:
            h = h * node_mask
        return h, x
```
https://github.com/ehoogeboom/e3_diffusion_for_molecules/blob/fce07d701a2d2340f3522df588832c2c0f7e044a/egnn/egnn_new.py#L134C5-L134C5

2) Compared to what's explained in section B of the supplementary information of the paper, phi_x has an extra `tanh()` function and its output is rescaled by a constant (`self.coords_range`). This can be seen in the `coord_model` method of the `EquivariantUpdate` class.
```
    def coord_model(self, h, coord, edge_index, coord_diff, edge_attr, edge_mask):
        row, col = edge_index
        input_tensor = torch.cat([h[row], h[col], edge_attr], dim=1)
        if self.tanh:
            trans = coord_diff * torch.tanh(self.coord_mlp(input_tensor)) * self.coords_range
        else:
            trans = coord_diff * self.coord_mlp(input_tensor)
        if edge_mask is not None:
            trans = trans * edge_mask
        agg = unsorted_segment_sum(trans, row, num_segments=coord.size(0),
                                   normalization_factor=self.normalization_factor,
                                   aggregation_method=self.aggregation_method)
        coord = coord + agg
        return coord
```
https://github.com/ehoogeboom/e3_diffusion_for_molecules/blob/fce07d701a2d2340f3522df588832c2c0f7e044a/egnn/egnn_new.py#L86C1-L99C21

The default value for `self.tanh` is `True`, which is inherited by the default `--tanh` argument in `main_qm9.py`. This can be simplify fixed by setting the default to `False`.

3) The code uses gradient clipping with a moving-averaged maximal norm, which is not mentioned at all in the paper.
This can be seen in the `train_epoch` function in `train_test.py`:
```
        if args.clip_grad:
            grad_norm = utils.gradient_clipping(model, gradnorm_queue)
        else:
            grad_norm = 0.
```
https://github.com/ehoogeboom/e3_diffusion_for_molecules/blob/fce07d701a2d2340f3522df588832c2c0f7e044a/train_test.py#L59C1-L62C27

The default value for `args.clip_grad` is `True` in `main_qm9.py`. This could also be fixed by setting the default, but I now wonder if this was used for producing the results of the paper. Some clarifications would be appreciated.

**Comment by ehoogeboom** [comment]
Hopefully the code should be correct. If I remember correctly the `egnn_new.py` is a refactor so you could compare to the functions/classes from eggn.py. From what I can see, points 1/2/3 in code is how they were used.
